core/arch/riscv/mm/core_mmu_arch.c

```diff
@@ -0,0 +1,724 @@
+		if (entry & PTE_W)
+			mattr |= TEE_MATTR_PW;
+		if (entry & PTE_X)
+			mattr |= TEE_MATTR_PX;
+	}
+
+	if (entry & PTE_G)
+		mattr |= TEE_MATTR_GLOBAL;
+
+	return mattr;
+}
+
+static uint8_t mattr_to_perms(unsigned level __maybe_unused,
+			      uint32_t attr)
+{
+	unsigned long perms = 0;
+
+	if (attr & TEE_MATTR_TABLE)
+		return PTE_V;
+
+	if (attr & TEE_MATTR_VALID_BLOCK)
+		perms |= PTE_V;
+
+	if (attr & TEE_MATTR_UR)
+		perms |= PTE_R | PTE_U;
+	if (attr & TEE_MATTR_UW)
+		perms |= PTE_W | PTE_U;
+	if (attr & TEE_MATTR_UX)
+		perms |= PTE_X | PTE_U;
+
+	if (attr & TEE_MATTR_PR)
+		perms |= PTE_R;
+	if (attr & TEE_MATTR_PW)
+		perms |= PTE_W | PTE_R;
+	if (attr & TEE_MATTR_PX)
+		perms |= PTE_X | PTE_R;
+
+	if (attr & TEE_MATTR_GLOBAL)
+		perms |= PTE_G;
+
+	return perms;
+}
+
+static unsigned int core_mmu_pgt_idx(vaddr_t va, unsigned int level)
+{
+	unsigned int idx = va >> CORE_MMU_SHIFT_OF_LEVEL(level);
+
+	return idx & RISCV_MMU_VPN_MASK;
+}
+
+static struct mmu_partition *core_mmu_get_prtn(void)
+{
+	return &default_partition;
+}
+
+static struct mmu_pgt *core_mmu_get_root_pgt_va(struct mmu_partition *prtn)
+{
+	return prtn->root_pgt;
+}
+
+static struct mmu_pgt *core_mmu_get_ta_pgt_va(struct mmu_partition *prtn)
+{
+	return &prtn->user_pgts[thread_get_id()];
+}
+
+static struct mmu_pgt *core_mmu_pgt_alloc(struct mmu_partition *prtn)
+{
+	struct mmu_pgt *pgt = NULL;
+
+	if (prtn->pgts_used >= RISCV_MMU_MAX_PGTS) {
+		debug_print("%u pgts exhausted", RISCV_MMU_MAX_PGTS);
+		panic();
+		return NULL;
+	}
+
+	pgt = &prtn->pool_pgts[prtn->pgts_used++];
+
+	memset(pgt, 0, RISCV_MMU_PGT_SIZE);
+
+	debug_print("pgts used %u / %u", prtn->pgts_used, RISCV_MMU_MAX_PGTS);
+
+	return pgt;
+}
+
+static void core_init_mmu_prtn_ta_core(struct mmu_partition *prtn __unused,
+				       unsigned int core __unused)
+{
+	/*
+	 * user_va_idx is the index in CORE_MMU_BASE_TABLE_LEVEL.
+	 * The entry holds pointer to the user mapping table in next level
+	 * that changes per core. Therefore, nothing to do.
+	 */
+}
+
+static void core_init_mmu_prtn_ta(struct mmu_partition *prtn)
+{
+	unsigned int core = 0;
+
+	assert(user_va_idx != -1);
+
+	memset(prtn->user_pgts, 0, CFG_NUM_THREADS * RISCV_MMU_PGT_SIZE);
+	for (core = 0; core < CFG_TEE_CORE_NB_CORE; core++)
+		core_init_mmu_prtn_ta_core(prtn, core);
+}
+
+static void core_init_mmu_prtn_tee(struct mmu_partition *prtn,
+				   struct tee_mmap_region *mm)
+{
+	size_t n = 0;
+	void *pgt = core_mmu_get_root_pgt_va(prtn);
+
+	memset(pgt, 0, RISCV_MMU_PGT_SIZE);
+	memset(prtn->pool_pgts, 0, RISCV_MMU_MAX_PGTS * RISCV_MMU_PGT_SIZE);
+
+	for (n = 0; !core_mmap_is_end_of_table(mm + n); n++)
+		if (!core_mmu_is_dynamic_vaspace(mm + n))
+			core_mmu_map_region(prtn, mm + n);
+}
+
+void tlbi_va_range(vaddr_t va, size_t len,
+		   size_t granule)
+{
+	assert(granule == CORE_MMU_PGDIR_SIZE || granule == SMALL_PAGE_SIZE);
+	assert(!(va & (granule - 1)) && !(len & (granule - 1)));
+
+	/* memory barrier */
```

Comment:
It would be more useful to comment on why this memory barrier is needed. Same below in this function and in the next.